### PR TITLE
buffer: separate the BufferFragement release and drain tracker

### LINKED ISSUE #28760: Envoy 1.27.0: Regression in UserSpace::IoHandleImpl
*Title*: New `reset_drain_trackers_and_accounting functionality` on `Buffer::Instance::move` caused a regression in Envoy 1.27.0

*Description*:
The logic for `moveUpTo` in `io_handle_impl.cc` has changed this line:
```
dst.move(src, res);
```
.. to this:
```
dst.move(src, res, /*reset_drain_trackers_and_accounting=*/true);
```
.. as part of this PR: https://github.com/envoyproxy/envoy/pull/27499

The PR claims the following:
_"we don't care about the write buffer as soon as it has been written to the peer's IO handle."_

.. but `UserSpace::IoHandleImpl` implements the "write" by moving the buffer slice to another buffer instance **without** expecting the slice to be marked as invalid.  The problem is that the drainer on `Buffer::Slice` is set up in the constructor as:
```
addDrainTracker([&fragment]() { fragment.done(); });
```
.. and `fragment.done()` invalidates the slice per the following comment for `BufferFragment` in `buffer.h`:
```
  /**
   * Called by a buffer when the referenced data is no longer needed.
   */
  virtual void done() PURE;
```

The end result is that slices sent using the userspace IO handle are invalidated before being moved to the other buffer.

Snippet from call stack:
```
#13 0x0000000006533cd9 in Envoy::Buffer::Slice::Slice(Envoy::Buffer::BufferFragment&)::{lambda()#1}::operator()() const (this=0x48f7edd0d10) at external/envoy/source/common/buffer/buffer_impl.h:93
 :
#16 0x0000000002632856 in Envoy::Buffer::Slice::callAndClearDrainTrackersAndCharges (this=0x48f7ee9d0d8) at external/envoy/source/common/buffer/buffer_impl.h:325
#17 0x000000000652878f in Envoy::Buffer::OwnedImpl::move (this=0x48f7ec94aa8, rhs=..., length=49595, reset_drain_trackers_and_accounting=true) at external/envoy/source/common/buffer/buffer_impl.cc:351
#18 0x0000000005933688 in Envoy::Buffer::WatermarkBuffer::move (this=0x48f7ec94aa8, rhs=..., length=49742, reset_drain_trackers_and_accounting=true) at external/envoy/source/common/buffer/watermark_buffer.cc:72
#19 0x0000000004af8635 in Envoy::Extensions::IoSocket::UserSpace::(anonymous namespace)::moveUpTo (dst=..., src=..., max_length=131072) at external/envoy/source/extensions/io_socket/user_space/io_handle_impl.cc:46
#20 0x0000000004af9cdf in Envoy::Extensions::IoSocket::UserSpace::IoHandleImpl::write (this=0x48f7ec94e00, buffer=...) at external/envoy/source/extensions/io_socket/user_space/io_handle_impl.cc:276
#21 0x0000000005a1a593 in Envoy::Network::RawBufferSocket::doWrite (this=0x48f7ec5ff20, buffer=..., end_stream=false) at external/envoy/source/common/network/raw_buffer_socket.cc:64
#22 0x000000000591a586 in Envoy::Network::ConnectionImpl::onWriteReady (this=0x48f7f1eea80) at external/envoy/source/common/network/connection_impl.cc:712
```

*Repro steps*:
I cannot share my application source, but it is running a test case that sends a large payload in multiple 4K slices through the internal listener (userspace IO handle) mechanism.  The test succeeds with versions of Envoy prior to 1.27.0.

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: buffer: separate the BufferFragement release and drain tracker
Additional Description:
There is a requirement for the user space io handle and iouring io handle to move the write buffer to its own buffer and invoke the drain trackers. But becomes an issue for the BufferFragment, the BufferFragement's releasor is invoked by the drain tracker also. After moving with invoke drain trackers, the memory in the BufferFragment will be released too, which leads to the Buffer reference some memory already released.
Risk Level: high
Testing: unittest
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Fixes #28760
Related to issue: https://github.com/envoyproxy/envoy/issues/28395

### CODE REVIEW

**Comment by soulxu** [comment]
/wait, I want to give a test with IoUring first.

**Comment by soulxu** [comment]
@KBaichoo @yanavlasov This should be ready for the review, I also tested with https://github.com/envoyproxy/envoy/pull/28824, it didn't see any fail related to use the BufferFragment.

File: source/common/buffer/buffer_impl.h [hunk 1]
```diff
@@ -307,6 +314,7 @@ class Slice {
   void transferDrainTrackersTo(Slice& destination) {
     destination.drain_trackers_.splice(destination.drain_trackers_.end(), drain_trackers_);
     ASSERT(drain_trackers_.empty());
+    destination.releasor_.swap(releasor_);
```
Comment (KBaichoo) [inline source/common/buffer/buffer_impl.h:317]: what happens with the original releasor that the destination has? Is it fine to toss it as we're doing? ISTM that we don't need to do this especially as the only place `transferDrainTrackersTo` is called is when we copy the current slice into destination in `coalesceOrAddSlice`.

**Review by KBaichoo — COMMENTED** [review]
/wait

would be awesome if we could have a reproduction in an integration test similar to what @roelfdutoit  reported in https://github.com/envoyproxy/envoy/issues/28760

cc @pianiststickman since this seems like a regression related to your change in #27499

File: source/common/buffer/buffer_impl.h (hunk 1 above)
Comment (soulxu) [reply source/common/buffer/buffer_impl.h:317]: yes, you are right, I shouldn't change releasor here.

**Comment by soulxu** [comment]
> would be awesome if we could have a reproduction in an integration test similar to what @roelfdutoit reported in #28760
> 
I will look at how can I reproduce that in an integration test

**Comment by soulxu** [comment]
@KBaichoo I added the tests for UserSpace::IoHandle, would you like take a look this again? thanks in advance!

File: source/common/buffer/buffer_impl.h [hunk 2]
```diff
@@ -307,6 +314,9 @@ class Slice {
   void transferDrainTrackersTo(Slice& destination) {
     destination.drain_trackers_.splice(destination.drain_trackers_.end(), drain_trackers_);
     ASSERT(drain_trackers_.empty());
+    // The releasor needn't to be transferred, and actually if there is releasor, this
+    // slice can't coalesce. Then there won't be a chance to calling this method.
+    ASSERT(releasor_ == nullptr);
```
Comment (KBaichoo) [inline source/common/buffer/buffer_impl.h:319]: I might be missing something on my end but how can we guarantee no slice that we're transferring (copying) has no releasor?

**Review by KBaichoo — COMMENTED** [review]
Thanks for working on this

/wait

File: source/common/buffer/buffer_impl.h (hunk 2 above)
Comment (soulxu) [reply source/common/buffer/buffer_impl.h:319]: There is condition control that https://github.com/envoyproxy/envoy/blob/782bb0c102ffbd9880615526c54c00301a6f3f82/source/common/buffer/buffer_impl.cc#L298-L303

The `other_slice.canCoalesce()` is false for the Slice created from BufferFragement.
https://github.com/envoyproxy/envoy/blob/782bb0c102ffbd9880615526c54c00301a6f3f82/source/common/buffer/buffer_impl.h#L89-L90

Comment (KBaichoo) [reply source/common/buffer/buffer_impl.h:319]: That makes sense, thanks @soulxu

**Review by KBaichoo — APPROVED** [review]
/assign @ggreenway 

For additional buffer expertise

**Comment by soulxu** [comment]
@ggreenway gentle ping :)

File: source/common/buffer/buffer_impl.h [hunk 3]
```diff
@@ -119,6 +120,7 @@ class Slice {
       reservable_ = rhs.reservable_;
       drain_trackers_ = std::move(rhs.drain_trackers_);
       account_ = std::move(rhs.account_);
+      releasor_.swap(rhs.releasor_);
```
Comment (ggreenway) [inline source/common/buffer/buffer_impl.h:123]: I don't think `swap` is correct in this case. Putting the old releasor into `rhs` is incorrect, isn't it?

**Review by ggreenway — CHANGES_REQUESTED** [review]
Sorry for the delay on this. It was marked `waiting` so it didn't show up my list.

/wait

File: source/common/buffer/buffer_impl.h (hunk 3 above)
Comment (soulxu) [reply source/common/buffer/buffer_impl.h:123]: You are right, let me fix that

**Comment by soulxu** [comment]
> Sorry for the delay on this. It was marked `waiting` so it didn't show up my list.
> 
> /wait

no worries! I didn't pay attention on this also

**Review by ggreenway — APPROVED** [review]

**Comment by soulxu** [comment]
It seems stuck at DCO test.

**Comment by soulxu** [comment]
@ggreenway ci is passed, would you think it is ok to merge?

**Pull request merged**